### DISCUSSION TITLE
why do we clamp span dimensions to a cellHeight grid

### DISCUSSION THREAD

**Discussion by TrippTrapp84** [recastnavigation/recastnavigation#739] [Q&A]
I've been digging through the source to learn how recast works, and on lines 444-447 of the file `ReacstRasterization.cpp` I see this:
```cpp
unsigned short spanMinCellIndex = (unsigned short)rcClamp((int)floorf(spanMin * inverseCellHeight), 0, RC_SPAN_MAX_HEIGHT);
unsigned short spanMaxCellIndex = (unsigned short)rcClamp((int)ceilf(spanMax * inverseCellHeight), (int)spanMinCellIndex + 1, RC_SPAN_MAX_HEIGHT);

if (!addSpan(heightfield, x, z, spanMinCellIndex, spanMaxCellIndex, areaID, flagMergeThreshold)) {
```

As far as I understand, these values `spanMinCellIndex` and `spanMaxCellIndex` are used as the min and max bounds that this span  occupies in a single large vertical column on our XZ heightfield. If that understanding is correct, can someone explain the reason for rounding the `spanMin` and `spanMax` to the nearest cellHeight like this? To me it seems as if this looses precision by rounding the heights of the span, no? I'm sure my understanding is just wrong somewhere and if anyone could explain in some detail the rationale and reasons for this, it would greatly help!

**Comment by rob-myers** [comment]
Maybe these are voxel units? 

> Units are usually in voxels (vx) or world units (wu). The units for voxels, grid size, and cell size are all based on the values of [cs](https://recastnav.com/structrcConfig.html#ae74c4bd2b0477f29a65a53ca9c59196d) and [ch](https://recastnav.com/structrcConfig.html#acdde0dfdbb39611966ece78ef473d861).

**Reply by TrippTrapp84** [reply]
I figure it probably is something like that. But from what I saw in the source, it looked like we never use it as a voxel index. The column spans are a linked list in our heightfield, which is itself 2-dimensional (From what I understand).

Maybe these values are used later in the code as a voxel index? I'm not sure though.

**Reply by rob-myers** [reply]
I'm no expert, I would like to understand the code better too. 

Apparently a 3d voxel representation is created, and afterwards it is smoothed. There are some nice diagrams here:

> https://www.unrealdoc.com/p/navigation-mesh
